File: src/main/java/com/rednorte/sigle/pacientes_service/listener/CancelacionListener.java
package com.rednorte.sigle.pacientes_service.listener;

import com.rednorte.sigle.pacientes_service.model.Notificacion;
import com.rednorte.sigle.pacientes_service.repository.NotificacionRepository;

import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;

@Component
@RequiredArgsConstructor
@Slf4j
public class CancelacionListener {

    private final NotificacionRepository notificacionRepository;

    @RabbitListener(queues = "sigle.citas.canceladas")
    public void recibirCancelacion(CancelacionEvento evento) {
        log.info("Evento de cancelación recibido para la cita {} del paciente {}",
                evento.getCitaId(), evento.getPacienteId());

        String mensaje = String.format(
            "Estimado paciente, lamentamos informar que su cita ha sido cancelada por el siguiente motivo: %s. Pronto será reasignado.",
            evento.getMotivo()
        );

        Notificacion notificacion = Notificacion.builder()
                .pacienteId(evento.getPacienteId())
                .tipo("CANCELACION_CITA")
                .canal("EMAIL")
                .mensaje(mensaje)
                .estado("ENVIADA")
                .enviadoEn(LocalDateTime.now())
                .eventoOrigen("CitasService.Cancelacion")
                .build();

        notificacionRepository.save(notificacion);
        log.info("Notificación guardada/enviada con éxito.");
    }

    @Data
    public static class CancelacionEvento {
        private Long citaId;
        private Long pacienteId;
        private Long listaEsperaId;
        private String motivo;
    }
}

File: src/listeners/cancelacionListener.js
const amqp = require('amqplib');
const service = require('../services/notificacionService');

const EXCHANGE = 'sigle.exchange';
const ROUTING_KEY = 'citas.cancelada';
const QUEUE = 'sigle.citas.canceladas';

async function startListener() {

    try {

        const conn = await amqp.connect(process.env.RABBITMQ_URL);

        const channel = await conn.createChannel();

        await channel.assertExchange(EXCHANGE, 'direct', {
            durable: true
        });

        await channel.assertQueue(QUEUE, {
            durable: true
        });

        await channel.bindQueue(
            QUEUE,
            EXCHANGE,
            ROUTING_KEY
        );

        console.log(`[RabbitMQ] Escuchando cola: ${QUEUE}`);

        channel.consume(QUEUE, async (msg) => {

            console.log("========== CANCELACIÓN RECIBIDA ==========");

            if (!msg) return;

            console.log(msg.content.toString());

            try {

                const evento = JSON.parse(msg.content.toString());

                const mensaje =
                    `Estimado paciente, lamentamos informar que su cita ha sido cancelada por el siguiente motivo: ${evento.motivo}. Pronto será reasignado.`;

                await service.create({
                    pacienteId: evento.pacienteId,
                    tipo: 'CANCELACION_CITA',
                    canal: 'EMAIL',
                    mensaje,
                    eventoOrigen: 'CitasService.Cancelacion',
                });

                console.log("[RabbitMQ] Notificación de cancelación creada.");

                channel.ack(msg);

            } catch (err) {

                console.error("[RabbitMQ] Error:", err);

                channel.nack(msg, false, false);

            }

        });

        conn.on('close', () => {
            console.warn("[RabbitMQ] Conexión cerrada.");
            setTimeout(startListener, 5000);
        });

    } catch (err) {

        console.error("[RabbitMQ] Error:", err);

        setTimeout(startListener, 5000);

    }

}

module.exports = {
    startListener
};